### ISSUE TITLE
[Bug] Runtime API Call Decoding incorrect with Custom Types

### ISSUE THREAD

**Issue by wischli** [polkadot-js/api#5298]
* **I'm submitting a ...**

  - [x] Bug report
  - [ ] Feature request
  - [ ] Support request
  - [ ] Other

* **What is the current behavior and expected behavior?**

*Current*: In our case, calling a custom RPC and its runtime API clone with the same input value does not result in the same return value: One field is decoded incorrectly. The RPC returns the right value and encodes the amount to a number, while the runtime API returns a different account (probably a double encoding), and a HEX-encoded value for the deposit amount.

```rust
struct Deposit {
   owner: AccountId,
   amount: u128
}
```

❌ Runtime Api: `didApi.queryByW3n("williamfreude")`
```
owner: 4nv4phaiCRnr6JSfhawcDZp7LzmvtWvabyCxkML7NPrr8eHV
amount: 8,589,934,592,000,000,154,110,983
```

✅ DID RPC: `did.queryByW3n("williamfreude")`
```
owner: 4qTuS5ymocwsPJ3uK5FKG7d2nVhV3CawDMwhWyKz2FBXq2gX
amount: 2,000,000,000,000,000
```

We tried out numerous varations, with and without custom `serde` (de-) serialization. Nothing worked until we used minimal custom type declarations, see below.

*Expected*

When calling the exact same method with the same input and same custom type declarations, it should lead to the same result. 

We fixed the issue for the runtime API call by referring to augmented api types, see [this PR for changing our custom type declaration](https://github.com/KILTprotocol/sdk-js/pull/670/files#diff-26c4d60dee56f71343deb202393345d1c8436e1cb4deb6f1c842fbcf3ff36646R18)
However, this augmented type definition matches exactly the custom one which we provided:

[Augmented](https://github.com/KILTprotocol/sdk-js/blob/develop/packages/augment-api/src/interfaces/types-lookup.ts#L3667)
```ts
  interface DidDidDetails extends Struct {
    readonly authenticationKey: H256;
    readonly keyAgreementKeys: BTreeSet<H256>;
    readonly delegationKey: Option<H256>;
    readonly attestationKey: Option<H256>;
    readonly publicKeys: BTreeMap<H256, DidDidDetailsDidPublicKeyDetails>;
    readonly lastTxCounter: u64;
    readonly deposit: KiltSupportDeposit;
  }
```

[Custom](https://github.com/KILTprotocol/sdk-js/pull/670/files#diff-26c4d60dee56f71343deb202393345d1c8436e1cb4deb6f1c842fbcf3ff36646L14)
```ts
export interface RawDidDetails extends Struct {
  readonly authenticationKey: Hash;
  readonly keyAgreementKeys: BTreeSet<Hash>;
  readonly delegationKey: Option<Hash>;
  readonly attestationKey: Option<Hash>;
  readonly publicKeys: BTreeMap<Hash, DidDidDetailsDidPublicKeyDetails>;
  readonly lastTxCounter: BlockNumber;
  readonly deposit: KiltSupportDeposit;
}
```
* **What is the motivation for changing the behavior?**

Others shouldn't need to spend time debugging this unexpected behavior in the future.

* **Please tell us about your environment:**

I tested both in the Polkadot Apps as well as a custom (Type)script.

  - Version: `9.4.1` and `9.6.1`
  - Environment:

    - [x] Node.js (`node v16.13.2 (npm v8.3.1)`)
    - [x] Browser (Polkadot Apps `apps v0.121.2-118-x` and Chrome `106.0.5249.119`. Both also with earlier versions, ca. 2 week range)
    - [ ] Other (limited support for other environments)

  - Language:

    - [ ] JavaScript
    - [x] TypeScript `4.8.3`
    - [x] Other: Polkadot Apps

**Comment by jacogr** [comment]
You need to link to -

- the Rust code implementing the state call
- the injected definitions for the JS

**Comment by wischli** [comment]
> You need to link to -
> 
> * the Rust code implementing the state call and RPC

https://github.com/KILTprotocol/kilt-node/blob/master/runtimes/standalone/src/lib.rs#L875

> * the injected definitions for the JS (runtime and RPC)

* [Runtime Call](https://github.com/KILTprotocol/type-definitions/blob/master/src/runtime.ts#L7) returns `RawDidLinkedInfo`
* [RPC Calls](https://github.com/KILTprotocol/type-definitions/blob/master/src/rpc.ts#L21) returns `DidLinkedInfo`

However, both types are equal on [the rust runtime](https://github.com/KILTprotocol/kilt-node/blob/master/rpc/did/runtime-api/src/lib.rs#L89) and our [type decorations](https://github.com/KILTprotocol/type-definitions/blob/master/src/types_10720.ts#L5) were picked from the augmented API types. Based on your comment, the root of our issue should be here by not having 1-1 same type decorations. Thank you!

**Comment by woss** [comment]
i'll be keeping this active as long as it is not resolved :)

**Comment by woss** [comment]
i'll be keeping this active as long as it is not resolved :)

**Comment by woss** [comment]
i'll be keeping this active as long as it is not resolved :)

**Comment by sinkingsugar** [comment]
What's the point of substrate if custom apis don't work (JS side, which is the 90% or use cases)... shocked to see this issue still open... everyone still using old deprecated RPCs? Or?

**Comment by jacogr** [comment]
If you are "shocked to see issues open", please make a PR. 

This is the reality -

1. There is 1 of me
2. There are a literal shit ton of repos to look at for me - pick anything in the polkadot-js org, that is all me
3. Each of the repos you pick can actually do with a team - read the first point.
4. If you want a quick resolution - dive in and make a PR

So please stop bitching. I get to stuff when I get there and can focus.

**Comment by jacogr** [comment]
@wischli I may have some time in the next week to take an initial look - however the links you posted are non-permanent so doesn’t quite point to the actual code (anymore).

**Comment by woss** [comment]
>This is the reality -
> 1. There is 1 of me

@jacogr  that's a sad reality, especially since you are still doing all these things alone. How come the parity or w3f team didn't hire more people to help you? I'm sure there are people out there with skills and in the need of work.

**Comment by jacogr** [comment]
Let's focus on the topic at-hand after my little outbursts :)

**Comment by sinkingsugar** [comment]
> If you are "shocked to see issues open", please make a PR.
> 
> This is the reality -
> 
> 1. There is 1 of me
> 2. There are a literal shit ton of repos to look at for me - pick anything in the polkadot-js org, that is all me
> 3. Each of the repos you pick can actually do with a team - read the first point: there is ONE person for all of this.
> 4. If you want a quick resolution - dive in and make a PR
> 
> TL;DR So please stop bitching. I get to stuff when I get there and can focus, no sooner. If it doesn't align with your timelines, stop complaining and get coding.

Isn't parity backing you? That was my assumption.
Do you need support?

**Comment by wischli** [comment]
Thank you for taking the time to look into this, especially given the facts stated above. You are doing a tremendous job, kudos to you!

> TL;DR Looking at the struct above, it make sense why it fails on `lastTxCounter: 'BlockNumber'` and works on `lastTxCounter: 'u64'`, so would say that it operates as expected with the supplied definitions, i.e. yielding invalid results when incorrect.

How can that be possible if the type [has always been](https://github.com/KILTprotocol/kilt-node/blame/8ccb0c77afc25d9d10eb57487730a44881a01161/runtimes/common/src/lib.rs#L74) `u64` on the runtime? P.S. This time permalink, sorry about that 😅

**Comment by jacogr** [comment]
KILT Spiritnet -

![image](https://user-images.githubusercontent.com/1424473/224909600-e96bca49-c11d-46a1-812b-e0c69915e41d.png)

RILT -

![image](https://user-images.githubusercontent.com/1424473/224909758-6321476e-50f2-4d2e-8cc3-6953a1164609.png)

I can only show what it knows about - it knows that `BlockNumber` is `u32` in the above cases.

**Comment by jacogr** [comment]
Only runtimes 9 & 10 have `BlockNumber: 'u64'` overrides in https://github.com/KILTprotocol/sdk-js/tree/develop/packages/type-definitions/src

(Latest, looking at the imports, never apply any such override)

**Comment by ntn-x2** [comment]
Hi @jacogr, kudos also from me for the job you are doing!

Anyway, as far as I have seen, we can re-use as many runtime-declared types as we need for the runtime API, correct? This means that if we define a type as having a `BlockNumber`, the definition should be taken from the runtime. And since nothing else that involves a block number fails decoding, I am, or was, assuming the type declaration for a block number was correct.

So now I have two questions left at this point:

1. Is it correct to assume that there is a bug either in the generated metadata or in this library to properly define types for core components, such as block number, that deviate from the default ones? Because I don't see how using the default definition of `BlockNumber` would lead to a different underlying type than what the runtime defines.
2. If indeed the definition of `BlockNumber` is wrong, how can it be that nothing else fails to decode nor decodes weirdly?

Thanks a lot for taking time to look into this!

**Comment by jacogr** [comment]
There is no magic I can apply to extract `BlockNumber` - there are layers upon layers of types, each with injected trait configs that define a specific override for that type. Same with e.g. `Balance`, generally it is a type, when you get to `assets`, it is completely different.

So, no - this is not a buggy piece of shit software - there is just no way to magically extract everything since a definition would depend on where it is used. 

If a chain -

- uses metadata v14, and
- an extrinsic is called, or
- a state query is made, or
- a event is executed, then
- the types are fully specified, by metadata

For everything else, it is not.

(There are some types that magically do get detected, as far as possible, e.g. stuff like `MultiAddress`, `AccountId`, `ExtrinsicSignature` based on how [extrinsics are defined](https://github.com/polkadot-js/api/blob/4c3b6a660e345708f39d425589a989fe7535c8c1/packages/types/src/metadata/PortableRegistry/PortableRegistry.ts#L419-L431) - this also doesn't work on all chains, it depends on config, but _mostly_ detects as expected)

**Comment by jacogr** [comment]
Closing.

1. This is working exactly as expected with the supplied inputs
2. Runtime calls and RPC will always decode with the user-supplied types

If on point 2 you want changes - head over to the Substrate repo.

- Runtime call types will be exposed on metadata at some point (make a PR and contribute to get this done sooner if it is a major PITA for you)
- For RPCs types the Substrate developers decided against it, so it will never be
